### PR TITLE
Configurable retained turns after compaction

### LINKED ISSUE #202: configurable retained messages after compaction
Currently the auto-compaction and the /compact command replace the entire conversation with a single Compacted summary user message. This isn't the best for continuity. It might be better to keep a few of the last messages around so the agent can kind of see where we left off a bit better. 

# Acceptance Criteria
- new config option in `compaction` config section: `num_retained_turns`
- compaction sets all messages to inactive except x latest massages based on config setting
- tests

### PR DESCRIPTION
**Pull request by robot-fizzizist**
Closes #202

Add configurable `num_retained_turns` to the `[compaction]` config section. When set, compaction preserves the last N conversation turns after the summary instead of replacing the entire history, maintaining conversational continuity. Tool-use blocks are stripped from retained messages (original rows soft-deleted, cleaned versions re-inserted).

Implementation plan posted as a comment below.

### CODE REVIEW

**Comment by robot-fizzizist** [comment]
## Implementation Plan

### Problem Analysis

Currently, compaction replaces the **entire** conversation with a single `[Compacted]` summary user message. This breaks conversational continuity — the agent loses all recent context. The fix is to preserve the last N turns after compaction, giving the agent a "window" into recent exchanges.

### Deliverables

1. **`CompactionConfig.num_retained_turns`** — new config field (default `0`, preserving current behavior)
2. **`ConversationRepo::compact_retaining()`** — DB method that deactivates all rows, then inserts summary + cleaned retained messages as new active rows
3. **`compact_with()` update** — modified to accept `num_retained_turns`, only summarize messages being compacted (not retained ones), strip tool-use blocks from retained messages, and rebuild in-memory history correctly
4. **`Agent::compact()` update** — pass `num_retained_turns` from config through to `compact_with`
5. **Config template update** — add `num_retained_turns` to `[compaction]` section
6. **Tests** — unit tests across all layers

### Files to Create or Modify

| File | Change |
|------|--------|
| `src/config.rs` | Add `num_retained_turns: u32` to `CompactionConfig` (default `0`); update `CONFIG_TEMPLATE` |
| `src/session/conversation.rs` | Add `compact_retaining(summary, retain_count)` method; deactivates all rows then inserts summary + cleaned retained messages as new active rows |
| `src/agent/compact.rs` | Update `compact_with()` to accept `num_retained_turns`, modify prompt to only include compacted messages, strip tool-use from retained messages, rebuild history as `prefix + [summary] + retained_turns` |
| `src/agent/mod.rs` | Pass `num_retained_turns` from config through to `compact_with`; store field on `Agent` |

### Soft-Delete Strategy

No rows are ever hard-deleted. The approach:
1. **Deactivate all existing rows** (soft delete via `active = 0`)
2. **Insert the summary** as a new active row
3. **Insert cleaned versions** of retained messages as new active rows (tool-use, tool-result, thinking, and redacted-thinking blocks stripped — only text blocks preserved)
4. `load_history()` (`WHERE active = 1`) naturally picks up only the cleaned versions

This preserves original tool-use content in inactive rows for audit/recovery.

### Turn Definition

A "turn" is a user message plus its immediately following assistant message(s). Turns are counted from the end of the conversation. If the last message is a user message with no assistant reply yet, that partial turn is still retained.

### Acceptance Criteria

1. `num_retained_turns = 0` (default) produces identical behavior to the current implementation — full compaction with no retained turns
2. `num_retained_turns > 0` preserves the last N turns after the summary message, with tool-use blocks stripped from retained messages
3. If `num_retained_turns` exceeds the total number of persisted turns, compaction is a no-op (returns early with a message)
4. The compaction summary is generated from **only** the messages being compacted (not the retained ones)
5. In-memory history after compaction = `context_prefix + [Compacted summary] + retained_turns(stripped)`
6. DB state after compaction: all old rows inactive, summary + cleaned retained messages active
7. All existing tests continue to pass; new tests cover the new behavior

### Testing Approach

**`src/session/conversation.rs`** — new tests:
- `compact_retaining_deactivates_old_and_keeps_last_n_turns`: Insert 6 messages (3 turns), retain 2 turns → summary + 4 cleaned messages active, first turn inactive
- `compact_retaining_with_zero_is_equivalent_to_compact`: Verify `retain_count=0` behaves identically to existing `compact()`
- `compact_retaining_with_count_exceeding_total_is_noop`: If retain count ≥ total turns, return early without modifying DB
- `compact_retaining_on_single_turn_is_noop`: Can't compact a single turn

**`src/agent/compact.rs`** — new tests:
- `compact_with_retained_turns_strips_tool_use`: Verify that tool-use and tool-result blocks are stripped from retained messages, leaving only text blocks
- `compact_with_retained_turns_prompt_excludes_retained`: Verify the summarization prompt only includes messages being compacted
- `compact_with_zero_retained_turns_behaves_as_before`: Baseline regression

**`src/config.rs`** — verify default value and deserialization:
- `compaction_config_default_num_retained_turns_is_zero`

### Risks and Open Questions

1. **Empty retained messages**: After stripping tool-use blocks, a retained message could have zero content blocks. These should be filtered out entirely (not kept as empty messages).
2. **Re-compaction edge case**: If `num_retained_turns` is high relative to `max_context_window_len`, auto-compaction may re-trigger immediately after compacting. The existing `last_auto_compacted` guard prevents infinite loops but UX could be confusing. Accepted per issue discussion.

---
*Plan created by mach6*

**Comment by robot-fizzizist** [comment]
## Progress Update

Implemented `num_retained_turns` for configurable retained turns after compaction.

**Changes:**
- `CompactionConfig.num_retained_turns` field (default `0`, preserving current behavior) in `src/config.rs`
- `ConversationRepo::compact_retaining()` method in `src/session/conversation.rs` — deactivates all rows, inserts summary + cleaned retained messages as new active rows
- `compact_with()` updated in `src/agent/compact.rs` — accepts `num_retained_turns`, summarizes only compacted messages, strips non-text blocks from retained messages, rebuilds history as `prefix + [summary] + retained_turns`
- `Agent::compact()` passes `num_retained_turns` from config through to `compact_with`
- Config template updated with `num_retained_turns` in `[compaction]` section
- Snapshot tests updated for intro message rendering
- 841 tests passing, zero clippy warnings

**Commit:** `65631b1`

---
*Progress tracked by mach6*

**Comment by robot-fizzizist** [comment]
## Code Review

### Critical

None.

### Important

**1. Duplicated `count_turns` and `strip_non_text` logic between `compact.rs` and `conversation.rs`** (severity: high, confidence: 92)

Both `src/agent/compact.rs` and `src/session/conversation.rs` implement nearly identical `count_turns()` functions and `strip_non_text()`/inline-stripping logic. `compact.rs` has `count_turns`, `retained_start_index`, and `strip_non_text` as private functions. `conversation.rs` has `count_turns` and `retained_turns` (which combines `retained_start_index` + `strip_non_text`). These do the same turn-counting and text-filtering work independently. If either diverges, the in-memory history and the DB state will get out of sync. This should be extracted to a shared module or have `compact.rs` delegate to `conversation.rs`'s functions.

**2. Missing agent-level integration test for `compact_with` with retained turns** (severity: high, confidence: 90)

The PR adds logic to `compact_with` that rebuilds in-memory history as `prefix + summary + retained_messages` when `num_retained_turns > 0`. No test exercises this path end-to-end. The existing `compact_without_spawner_returns_error` test only checks the error case. A test using a mock `AgentSpawner` should verify the history is correctly rebuilt (prefix preserved, summary inserted, retained turns appended with tool-use blocks stripped).

**3. Potential in-memory/DB inconsistency when `compact_retaining` returns `Ok(false)` (no-op)** (severity: high, confidence: 85)

In `compact_with`, when `num_retained_turns > 0`, if `compact_retaining` returns `Ok(false)`, the code drops the session lock and returns the string `"Nothing to compact: all turns are retained."` However, the check for `compacted.is_empty()` happens *before* calling the sub-agent, while `compact_retaining` checks `retain_count >= total_turns` *after* loading from DB. Between these two checks, state could theoretically diverge (e.g., another process modifies the DB). More practically, the early-return path skips the normal `history.clear()` + rebuild that happens on success, leaving in-memory history in its pre-compaction state — which is correct, but the function's callers need to be aware that a "no-op" compaction does not touch in-memory history at all. This should be documented explicitly.

### Suggestions

**4. `compact_with` has significant code duplication between the `num_retained_turns == 0` and `> 0` branches** (severity: medium, confidence: 88)

The prompt-building logic (iterating messages, formatting role labels, matching content blocks) is nearly identical in both branches. The only differences are: (a) the `> 0` branch splits messages into compacted/retained, and (b) the `> 0` branch filters retained messages through `strip_non_text`. The prompt formatting could be extracted into a helper function to eliminate ~40 lines of duplication.

**5. Missing test for `with_compaction_config` propagating `num_retained_turns`** (severity: medium, confidence: 85)

The existing test `with_compaction_config_sets_max_context_window_len` only checks `max_context_window_len`. A parallel assertion for `num_retained_turns` should be added to verify the value flows through correctly.

**6. Missing agent-level test for early-return when all turns are retained** (severity: medium, confidence: 82)

The no-op path (`compact_retaining` returns `Ok(false)`) is only tested at the `conversation.rs` level. An agent-level test should verify the exact return value and that in-memory history is left untouched.

**7. `retained_turns` in `conversation.rs` duplicates `retained_start_index` + `strip_non_text` from `compact.rs`** (severity: medium, confidence: 80)

The `retained_turns` helper in `conversation.rs` reimplements turn-boundary detection and text-only filtering that already exists in `compact.rs`. This increases maintenance surface. Consider extracting shared helpers into `src/types.rs` or a dedicated `compact_helpers.rs`.

### Strengths

- Thorough test coverage at the DB/session level — 8 integration tests for `compact_retaining` covering the no-op case, zero-retained case, strip behavior, and empty-message dropping
- Clean separation of concerns: DB operations in `conversation.rs`, in-memory manipulation in `compact.rs`
- The soft-delete strategy preserves auditability
- Good edge case handling: empty messages after stripping are dropped, retain count exceeding total is a no-op
- Snapshot tests correctly updated

**Agents run:** code-reviewer, error-auditor, test-reviewer, completeness-checker

---
*Reviewed by mach6*

**Comment by robot-fizzizist** [comment]
## Review Assessment

https://github.com/Fizzizist/illustrious-manager/pull/203#issuecomment-4420817779

### Classifications

| Finding | Classification | Reasoning |
|---|---|---|
| 1. Duplicated `count_turns`/`strip_non_text` logic | **Genuine issue** | `count_turns` in `compact.rs` is a verbatim copy of `count_turns` in `conversation.rs`. `strip_non_text` duplicates the inline `filter_map` in `conversation.rs`'s `retained_turns`. If either diverges, in-memory and DB state will disagree. Extract to a shared module. |
| 2. Missing agent-level integration test for `compact_with` with retained turns | **Genuine issue** | No test calls `compact_with` or `Agent::compact()` with `num_retained_turns > 0`. Only an end-to-end agent test can verify both in-memory and DB views remain identical after compaction with retained turns. |
| 3. Potential in-memory/DB inconsistency on no-op | **False positive** | `compact_retaining` returns `Ok(false)` before touching any rows. The early return in `compact_with` skips the history rebuild, but since the DB is unchanged, in-memory and DB states remain consistent. No fix needed. |
| 4. Code duplication between `num_retained_turns == 0` and `> 0` branches | **Nitpick** | ~30 lines of prompt-building duplicated. DRY cleanup, not a correctness fix. Can be addressed in a follow-up. |
| 5. Missing test for `with_compaction_config` propagating `num_retained_turns` | **Nitpick** | Trivial field copy (`self.num_retained_turns = config.num_retained_turns`). Low risk of regression. |
| 6. Missing agent-level test for no-op early-return path | **Nitpick** | Covered at the DB level. Agent-level test adds marginal value. |
| 7. `retained_turns` in `conversation.rs` duplicates `compact.rs` logic | **Genuine issue** | Same root cause as Finding 1 — shared logic should be extracted. |

### Action Plan

1. **Extract `count_turns`, `retained_start_index`, and `strip_non_text` to a shared module** — Create a `src/agent/compact_helpers.rs` (or add to `src/types.rs`) with these functions as `pub(crate)`. Have both `compact.rs` and `conversation.rs` import from the shared location. This eliminates the duplication that is the root cause of Findings 1 and 7.

2. **Add an agent-level integration test for `compact_with` with retained turns** — Use a mock `AgentSpawner` (or `SequencedBackend` wrapped appropriately) to exercise the `compact_with` path with `num_retained_turns > 0`. Assert that in-memory history is `prefix + [summary] + retained_turns(stripped)` and that DB state matches.

---
*Assessment by mach6*

**Comment by robot-fizzizist** [comment]
## Progress Update — Review Fixes

Addressed findings from review:

**Finding 1/7 (Genuine): Duplicated logic** — Extracted `count_turns`, `retained_start_index`, `strip_non_text`, and `retained_turns` to a new shared module `src/compact_helpers.rs`. Both `compact.rs` and `conversation.rs` now import from the single source of truth.

**Finding 2 (Genuine): Missing agent-level integration test** — Added 3 integration tests in `agent/mod.rs`:
- `compact_with_retained_turns_rebuilds_history_correctly` — verifies DB state matches shared helpers, compacted messages absent
- `compact_with_retained_turns_strips_tool_use_from_retained` — verifies tool-use blocks stripped from retained messages
- `compact_with_zero_retained_turns_behaves_as_before` — regression baseline

**Finding 5 (Nitpick): Missing `num_retained_turns` assertion** — Added `num_retained_turns_for_test()` accessor and assertion in `with_compaction_config` test.

**Findings 3, 4, 6** — Assessed as false positive or nitpick; no changes needed.

**Commit:** `7692920`

---
*Progress tracked by mach6*

File: src/session/conversation.rs [hunk 1]
```diff
@@ -128,6 +129,67 @@ impl<'a> ConversationRepo<'a> {
...
+                let _ = self.session.conn.execute("ROLLBACK", ()).await;
+                return Err(e);
+            }
+        }
+
+        self.session
+            .conn
+            .execute("COMMIT", ())
+            .await
+            .context("Failed to commit compaction transaction")?;
```
Comment (Fizzizist) [inline src/session/conversation.rs]: Wrapping in a transaction should not be done manually like this. There should be a wrapper so that anything can be wrapped in a transaction.

File: src/session/conversation.rs [hunk 2]
```diff
@@ -128,6 +129,67 @@ impl<'a> ConversationRepo<'a> {
...
+    /// last message is a user message with no assistant reply yet, that partial
+    /// turn is still retained.
+    ///
+    /// Tool-use, tool-result, thinking, and redacted-thinking blocks are
+    /// stripped from retained messages — only text blocks are preserved.
+    /// Messages that become empty after stripping are dropped entirely.
+    ///
+    /// Returns `Ok(true)` if compaction was performed, or `Ok(false)` if it
+    /// was a no-op (retain_count >= total turns, meaning there's nothing to
+    /// compact).
```
Comment (Fizzizist) [inline src/session/conversation.rs]: No need for so much verbose commenting

File: src/agent/compact_helpers.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
+use crate::types::{ContentBlock, Message, Role};
+
+/// Count the number of turns in a message slice.
+///
+/// A "turn" starts with a user message and includes all immediately following
+/// assistant messages. Orphan assistant messages at the start form their own turn.
```
Comment (Fizzizist) [inline src/agent/compact_helpers.rs]: unecessary comment as well as the other comments in this file

**Review by Fizzizist — CHANGES_REQUESTED** [review]
I don't agree with the placement of `src/compact_helpers.rs` It should be either co-located or merged with the other compaction logic

**Comment by robot-fizzizist** [comment]
## Progress Update — Co-locate Helpers with Compaction Logic

Moved `compact_helpers.rs` from crate root (`src/`) to alongside the other compaction logic (`src/agent/compact_helpers.rs`), as requested in review. Both `compact.rs` and `conversation.rs` now import from `crate::agent::compact_helpers`. Module is `pub(crate)` so it's accessible across the crate without exposing publicly.

**Commit:** `cafdb1f`

---
*Progress tracked by mach6*

**Comment by robot-fizzizist** [comment]
## Progress Update — Review Fix Round 2

Addressed all three review comments:

1. **Transaction wrapper** — Extracted `begin_transaction()`, `commit_transaction()`, and `rollback_transaction()` helper methods on `ConversationRepo`. Both `compact()` and `compact_retaining()` now use these instead of inline `BEGIN TRANSACTION` / `ROLLBACK` / `COMMIT` calls. Rust's `AsyncFnOnce` is unstable so a closure-based wrapper isn't viable, but the helper methods achieve the same DRY goal.

2. **Verbose comments on `compact_retaining`** — Removed the 15-line doc comment, reduced to just the method signature.

3. **Unnecessary comments in `compact_helpers.rs`** — Removed all doc comments from `count_turns`, `retained_start_index`, `strip_non_text`, and `retained_turns`. The function and type names are self-documenting.

**Commit:** `c08d22b`

---
*Progress tracked by mach6*
